File: app/src/main/java/tpcreative/co/qrscanner/model/HistoryModel.java
package tpcreative.co.qrscanner.model;

import androidx.room.Ignore;

import com.google.zxing.BarcodeFormat;

import java.io.Serializable;
import java.util.UUID;

import tpcreative.co.qrscanner.common.Utils;

public class HistoryModel implements Serializable {
    public int id;
    public String email;
    public String subject ;
    public String message;
    public String phone;
    public double lat;
    public double lon;
    public String query;
    public String title;
    public String location;
    public String description;
    public String startEvent;
    public String endEvent;
    public long   startEventMilliseconds;
    public long   endEventMilliseconds;
    public String fullName;
    public String address;
    public String text;
    public String ssId;
    public boolean hidden;
    public String password;
    public String url;
    public String createType;
    public String networkEncryption;
    public String createDatetime;
    public String barcodeFormat;
    public boolean favorite;
    public String updatedDateTime;

    /*content_type_barcode*/
    public String contentUnique;
    public String contentUniqueForUpdatedTime;

    /*Custom fields*/
    public TypeCategories typeCategories;
    private boolean isChecked;
    private boolean isDeleted;

    public HistoryModel(){
        this.email = "";
        this.subject = "";
        this.message = "" ;
        this.phone = "";
        this.lat = 0;
        this.lon = 0;
        this.startEventMilliseconds = 0;
        this.endEventMilliseconds = 0;
        this.query = "";
        this.title = "";
        this.location = "";
        this.description = "";
        this.startEvent = "";
        this.endEvent = "";
        this.fullName = "";
        this.address = "";
        this.text = "";
        this.ssId = "";
        this.hidden = false;
        this.password = "";
        this.url = "";
        this.createType = "";
        this.networkEncryption = "";
        this.typeCategories = new TypeCategories();
        this.barcodeFormat = BarcodeFormat.QR_CODE.name();
        this.favorite = false;
        this.updatedDateTime = Utils.getCurrentDateTimeSort();
        this.contentUnique = "";
        this.contentUnique = "";
    }

    public HistoryModel(HistoryEntityModel item){
        this.id = item.id;
        this.email = item.email;
        this.subject = item.subject ;
        this.message = item.message;
        this.phone = item.phone;
        this.lat = item.lat;
        this.lon = item.lon;
        this.query = item.query;
        this.title = item.title;
        this.location = item.location;
        this.description = item.description;
        this.startEvent = item.startEvent;
        this.endEvent = item.endEvent;
        this.startEventMilliseconds = item.startEventMilliseconds;
        this.endEventMilliseconds = item.endEventMilliseconds;
        this.fullName = item.fullName;
        this.address = item.address;
        this.text = item.text;
        this.ssId = item.ssId;
        this.hidden = item.hidden;
        this.password = item.password;
        this.url = item.url;
        this.createType = item.createType;
        this.networkEncryption = item.networkEncryption;
        this.createDatetime = item.createDatetime;
        this.barcodeFormat = item.barcodeFormat;
        this.favorite = item.favorite;
        this.updatedDateTime = item.updatedDateTime;
        this.contentUnique = item.contentUnique;
        this.contentUniqueForUpdatedTime = item.contentUniqueForUpdatedTime;
    }

    public boolean isChecked() {
        return isChecked;
    }

    public void setChecked(boolean checked) {
        isChecked = checked;
    }

    public boolean isDeleted() {
        return isDeleted;
    }

    public void setDeleted(boolean deleted) {
        isDeleted = deleted;
    }

    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public String getUUId(){
        try {
            return UUID.randomUUID().toString();
        }
        catch (Exception e){
            return ""+System.currentTimeMillis();
        }
    }

    public int getCategoryId() {
        return typeCategories.getId();
    }

    public String getCategoryName() {
        return typeCategories.getType();
    }
}

File: app/src/main/java/tpcreative/co/qrscanner/model/HistoryModel.kt
package tpcreative.co.qrscanner.model
import com.google.zxing.BarcodeFormat
import tpcreative.co.qrscanner.common.Utils
import java.io.Serializable

class HistoryModel : Serializable {
    var id : Int? = 0
    var email: String?
    var subject: String?
    var message: String?
    var phone: String?
    var lat: Double?
    var lon: Double?
    var query: String?
    var title: String?
    var location: String?
    var description: String?
    var startEvent: String?
    var endEvent: String?
    var startEventMilliseconds: Long?
    var endEventMilliseconds: Long?
    var fullName: String?
    var address: String?
    var text: String?
    var ssId: String?
    var hidden: Boolean?
    var password: String?
    var url: String?
    var createType: String?
    var networkEncryption: String?
    var createDatetime: String? = null
    var barcodeFormat: String?
    var favorite: Boolean?
    var updatedDateTime: String?

    /*content_type_barcode*/
    var contentUnique: String?
    var contentUniqueForUpdatedTime: String?

    /*sync data*/
    var isSynced: Boolean?
    var uuId: String?

    /*Custom fields*/
    var typeCategories: TypeCategories? = null
    private var isChecked = false
    private var isDeleted = false

    constructor() {
        email = ""
        subject = ""
        message = ""
        phone = ""
        lat = 0.0
        lon = 0.0
        startEventMilliseconds = 0
        endEventMilliseconds = 0
        query = ""
        title = ""
        location = ""
        description = ""
        startEvent = ""
        endEvent = ""
        fullName = ""
        address = ""
        text = ""
        ssId = ""
        hidden = false
        password = ""
        url = ""
        createType = ""
        networkEncryption = ""
        typeCategories = TypeCategories()
        barcodeFormat = BarcodeFormat.QR_CODE.name
        favorite = false
        updatedDateTime = Utils.getCurrentDateTimeSort()
        contentUnique = ""
        contentUniqueForUpdatedTime = ""
        isSynced = false
        uuId = Utils.getUUId()
    }

    constructor(item: HistoryEntityModel?) {
        id = item?.id
        email = item?.email
        subject = item?.subject
        message = item?.message
        phone = item?.phone
        lat = item?.lat
        lon = item?.lon
        query = item?.query
        title = item?.title
        location = item?.location
        description = item?.description
        startEvent = item?.startEvent
        endEvent = item?.endEvent
        startEventMilliseconds = item?.startEventMilliseconds
        endEventMilliseconds = item?.endEventMilliseconds
        fullName = item?.fullName
        address = item?.address
        text = item?.text
        ssId = item?.ssId
        hidden = item?.hidden
        password = item?.password
        url = item?.url
        createType = item?.createType
        networkEncryption = item?.networkEncryption
        createDatetime = item?.createDatetime
        barcodeFormat = item?.barcodeFormat
        favorite = item?.favorite
        updatedDateTime = item?.updatedDateTime
        contentUnique = item?.contentUnique
        contentUniqueForUpdatedTime = item?.contentUniqueForUpdatedTime
        isSynced = item?.isSynced
        uuId = item?.uuId
    }

    constructor(item: HistoryModel?, isSynced: Boolean) {
        id = item?.id
        email = item?.email
        subject = item?.subject
        message = item?.message
        phone = item?.phone
        lat = item?.lat
        lon = item?.lon
        query = item?.query
        title = item?.title
        location = item?.location
        description = item?.description
        startEvent = item?.startEvent
        endEvent = item?.endEvent
        startEventMilliseconds = item?.startEventMilliseconds
        endEventMilliseconds = item?.endEventMilliseconds
        fullName = item?.fullName
        address = item?.address
        text = item?.text
        ssId = item?.ssId
        hidden = item?.hidden
        password = item?.password
        url = item?.url
        createType = item?.createType
        networkEncryption = item?.networkEncryption
        createDatetime = item?.createDatetime
        barcodeFormat = item?.barcodeFormat
        favorite = item?.favorite
        updatedDateTime = item?.updatedDateTime
        contentUnique = item?.contentUnique
        contentUniqueForUpdatedTime = item?.contentUniqueForUpdatedTime
        this.isSynced = isSynced
        uuId = item?.uuId
    }

    fun isChecked(): Boolean {
        return isChecked
    }

    fun setChecked(checked: Boolean) {
        isChecked = checked
    }

    fun isDeleted(): Boolean {
        return isDeleted
    }

    fun setDeleted(deleted: Boolean) {
        isDeleted = deleted
    }

    fun getId(): Int {
        return id ?: 0
    }

    fun setId(id: Int) {
        this.id = id
    }

    fun getCategoryId(): Int {
        return typeCategories?.getId() ?: 0
    }

    fun getCategoryName(): String? {
        return typeCategories?.getType()
    }
}